### ISSUE TITLE
introduce chainsimulator mode for development

### ISSUE THREAD

**Issue by marc0olo** [aeternity/aeternity#3692]
# Chainsimulator mode

## Marginal conditions
- Devtools need to be able to re-use existing APIs to submit transactions, etc.
- 1 node only setup (no peer connections expected)
   - no sync required, can be deactivated!
- Resource requirements should be decreased as much as possible. At least mining mustn't be required anymore for testing.

## Configuration
These params can be included in a new section `simulator` of the YAML or provided in the command line when executing the binary:

### Required
- `active` Used to activate the chainsimulator mode with a custom consensus algorithm. Exposes the chainsimulator management API if activated.
    - default: `false`
- `block_interval` Can be used to define a specific "mining interval" in seconds. By default blocks should be mined immediately.
    - default: `0` (immediately)
- `protocol_version` Define the protocol version to run the node with (`hard_forks` section should be omitted)
    - default: `5` (always the latest hardfork)
- `start_time` The unix timestamp to start with the first (or ongoing) block
    - default: the current unix timestamp
- `derive_from_mnemonic` Use a specific HD wallet mnemonic to generate initial accounts.
    - default: generated on-startup
-  `default_balance` The default account balance to fund the generated accounts with, specified in AE.
    - default: `500`
- `total_accounts` Number of accounts to generate at startup.
    - default: `10` 

### Optional (to be discussed)
- `db_path` Some users prefer to set a custom db-path
- `block_height_start` The height to start the creation of blocks at. Only relevant if the node is started using a snapshot.
    - default: the following block based on the latest block of the snapshot
- `impersonated_accounts` An array of accounts addresses to be "impersonated". We probably need to investigate how to achieve that combined with tooling (e.g. SDK).
    - default: `[]`

YAML example for the configuration:
```yaml
- simulator:
  - active: true
  - chain:
    - db_path: '/opt/custom_dev_snapshot' # optional
    - protocol_version: 5
    - block_interval: 0
    - start_time: 15299999292
    - fork_from: 'https://mainnet.aeternity.io' # optional
    - block_height_start: 444450 # optional
  - accounts:
    - derive_from_mnemonic: <mnemonic_here>
    - default_balance: 500
    - total_accounts: 10
    - impersonated_accounts: ['ak_...', 'ak_...', 'ak_...'] # optional 
```

#### Protocol version
Here the question is how the node should behave if it is started with a snapshot of the chain. I think it should use the latest version of the snapshot and then immediately mine a new  the protocol version defined in the config. If the protocol version provided is lower than the latest version in the snapshot the startup must fail.

#### Account funding
Regarding the account generation the question is if we want to keep the standard way like it is right now where we provide a genesis file that contains a JSON file with addresses and the corresponding amount to fund it. The above mentioned configuration options would be a bit more convenient for developers. In every case the "genesis" accounts and balances should be printed/logged. Maybe this doesn't matter at all if we decide to manage this via management API of the chainsimulator. This way the tooling could call the API after startup and provide the addresses and respective amounts.

#### Block Height Start
Not sure if this is relevant. Does anybody see a reason to introduce this?

#### Database path
Not sure if this can currently already be set. But independent from that I think the chainsimulator should - for simplicity - only support 1 specific db backend.

#### Fork from & impersonated accounts
This can be a very important feature if somebody want's to test functionality e.g. based on the current state of the mainnet. In best case we would be able to only fetch the top height from the remote node and be able continue the chain in the simulator.

On Ethereum there exist a lot of DeFi protocols like Compound or Aave which new projects might want to integrate into their decentralized applications. In some cases it is also required to impersonate other accounts - meaning that you need to be able to perform transactions on behalf of these accounts while not knowing the private key. I am currently not sure how tools & frameworks like Ganache or Hardhat are managing this. But they allow to broadcast transactions on behalf of these "impersonated" or "unlocked" accounts. We should definitely take a look into that.

See:
- https://hardhat.org/guides/mainnet-forking.html#customizing-hardhat-network-s-behavior
- https://www.youtube.com/watch?v=oBqS37ffNjE&t=292s

#### Start using environment variables
According to recent discussions the config options can also be provided via env-variables when executing the node's binary. So e.g. running `./aeternity start AE__SIMULATOR__ACTIVE=true` would start the chainsimulator with default values. Basically there are different options to run the node with the desired configuration, see https://docs.aeternity.io/en/stable/configuration/#user-provided-configuration. @nikita-fuchs Raise your voice if we need sth. else here. But I think we already have enough options and just need to make sure it can easily be used with out tooling that wraps around it 😛 

## Chainsimulator management API
This API should be active by default if node runs in chainsimulator mode and can be a simple http API provided in a new group/tag.

#### Update protocol version (trigger hardfork)
We need an endpoint that can be used to trigger a hardfork on-demand.

- PUT `/chainsimulator/protocol-version`
  - Request payload example: `{"value":  2}`

Should fail if the protocol version doesn't exist and if the protocol version is equal or lower than the current version.

#### Jump into the future by block or time
We need an endpoint to be able to jump into the future to test block-height or time-based dependent logic.

**By block**
- PUT `/chainsimulator/block-height`
  - Request payload example: `{"type": "delta", "value": 20}` (here we can distinguish between `delta` and actual `block` as already done for `TTL`)

Should fail if an actual block is provided which is lower than the current height.

**By time**
- PUT `/chainsimulator/time`
  - Request payload example `{"value": 86400}` (amount of seconds to move forward)

#### Fund accounts
It might be necessary to fund certain accounts on-demand where one would not want to rely on performing a `SpendTx` on behalf of an existing account. Not sure if it makes sense as we are able to fund accounts on startup already. Can probably be omitted.

#### Snapshot the chain state & revert to a specific chain state
This is a feature that Ganache provides. Currently wondering about the usecase. When would we need to make a snapshot during tests where we could revert back to at any point of time?

Probably the easiest way is just to allow the dev to jump back to a certain block height of the current chain by invoking the management API of the simulator.

### References
For Ethereum there exists Ganache where we can take a look at and pick what we need:
- https://github.com/trufflesuite/ganache-core#features
- https://github.com/trufflesuite/ganache-cli#options-1
- https://github.com/trufflesuite/ganache-cli#management-api-methods
- https://github.com/trufflesuite/ganache-cli#custom-methods

**Comment by marc0olo** [comment]
seems like naming could change as `devmode` is probably the better naming for this

**Comment by cna0** [comment]
Hello @marc0olo! I have to say, great proposal, very thorough!

Have you seen this issue? https://github.com/aeternity/aeternity/issues/3545
It's a simple maintenance/debug mode, but can it be useful for smart-contract developers also?

**Comment by marc0olo** [comment]
> Hello @marc0olo! I have to say, great proposal, very thorough!

Yeah smart contract devs really need this kind of functionality. Details can - of course - be changed if it makes sense.

> Have you seen this issue? #3545
> It's a simple maintenance/debug mode, but can it be useful for smart-contract developers also?

No I haven't seen it before. It seems like there is at least some overlap when it comes to deactivating sync e.g.. And we don't need to be able to switch to devmode during runtime (if this is required for maintenance mode). Also it's not only about deactivating mining and http endpoints. In case of the devmode we need to completely change the consensus and still allow transactions to be broadcasted via the http endpoints like usual.
